### ISSUE TITLE
pid1: when invoking generators, forward all their output to kmsg while waiting for them to finish

### ISSUE THREAD

**Issue by dustymabe** [systemd/systemd#15638]
**Is your feature request related to a problem? Please describe.**

Right now any errors in systemd generators go nowhere unless the generator is configured to send output to kmsg:

https://github.com/systemd/systemd/blob/f20078df0b568cf365eea8278e98170e59ce2b2d/man/systemd.generator.xml#L187-L193

So if your generator is in bash you would add something like this to the top of your script:

```
exec 1>/dev/kmsg
exec 2>&1
```

**Describe the solution you'd like**

Could we possibly just do this by default for generators? If the generator really didn't want any output going to kmsg then maybe they could explicitly send things to /dev/null instead. 

**Describe alternatives you've considered**

- The user adding redirection to `/dev/kmsg` in every generator.
- systemd *only* hooking up stdout/stderr of generators to kmsg when systemd debug logging is enabled instead of by default.

**Comment by poettering** [comment]
hmm, i am not sure we can safely do that. /dev/kmsg is not really suitable for stdout or stderr. It interprets `<n>` at the beginning of the line magically, and that's usually not what people expect from stderr...

**Comment by boucman** [comment]
yes but OTOH, I give a systemd training, and the standard expectation of people discovering the feature is that the  logs go... somewhere. Losing them is really counter-intuitive and they expect them to reach the journal eventually...

kmsg is not ideal, but it's the only tool we got and people do use "echo bla > /dev/kmsg" 

could systemd filter stdout/stderr for us, maybe ?

**Comment by dustymabe** [comment]
> hmm, i am not sure we can safely do that. /dev/kmsg is not really suitable for stdout or stderr. It interprets `<n>` at the beginning of the line magically, and that's usually not what people expect from stderr...

Interesting.. Would that be an argument that would support that we *should* do something here to somehow filter stdout/stderr and interface with `/dev/kmsg` appropriately rather than have users inappropriately write directly to `/dev/kmsg` as the docs suggest?

**Comment by cgwalters** [comment]
Low tech: change systemd to allocate a pipe for each one and capture + replay to journal at least the first <pipesize> bytes of logging.

Also it'd be nice if systemd generated units for the generators - for Fedora CoreOS our CI runs `systemctl --failed` as a sanity check, but that doesn't catch generators.  We could make each generator turn into a synthetic `$(basename generator).service` that wasn't actually executed as a unit, just appeared to be one.

**Comment by dustymabe** [comment]
Support for `systemctl --failed` would be real nice, since most people won't look at the console output unless they observe a problem. The symptom of the issue could take a long time to be discovered.

**Comment by poettering** [comment]
i guess PID 1 could invoke all generators with a pipe on stdout/stderr and pass that to kmsg on its own while it waits for the generators to finish.

**Comment by snarkmaster** [comment]
A possible workaround for those stuck debugging generators in the current world -- wrap your generator with this sort of shell script:

```
#!/bin/sh -x
2>/dev/kmsg && exec 2> /dev/kmsg
1>/dev/kmsg && exec 1> /dev/kmsg
exec /actual/generator "$@"
```
